File: src/ImpactVectorBorneDiseasesComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;
using Fund.CommonDimensions;

namespace Fund.Components
{
    public interface IImpactVectorBorneDiseasesState
    {
        IVariable2Dimensional<Timestep, Region, double> dengue { get; }
        IVariable2Dimensional<Timestep, Region, double> schisto { get; }
        IVariable2Dimensional<Timestep, Region, double> malaria { get; }

        IParameter1Dimensional<Region, double> dfbs { get; }
        IParameter1Dimensional<Region, double> dfch { get; }
        IParameter1Dimensional<Region, double> smbs { get; }
        IParameter1Dimensional<Region, double> smch { get; }
        IParameter1Dimensional<Region, double> malbs { get; }
        IParameter1Dimensional<Region, double> malch { get; }


        IParameter1Dimensional<Region, double> gdp90 { get; }
        IParameter1Dimensional<Region, double> pop90 { get; }

        IParameter2Dimensional<Timestep, Region, double> income { get; }
        IParameter2Dimensional<Timestep, Region, double> population { get; }
        IParameter2Dimensional<Timestep, Region, double> temp { get; }

        double dfnl { get; }
        double vbel { get; }
        double smnl { get; }
        double malnl { get; }
    }

    public class ImpactVectorBorneDiseasesComponent
    {
        public void Run(Clock clock, IImpactVectorBorneDiseasesState state, IDimensions dimensions)
        {
            var s = state;
            var t = clock.Current;

            foreach (var r in dimensions.GetValues<Region>())
            {
                double ypc = 1000.0 * s.income[t, r] / s.population[t, r];
                double ypc90 = s.gdp90[r] / s.pop90[r] * 1000.0;

                s.dengue[t, r] = s.dfbs[r] * s.population[t, r] * s.dfch[r] * Math.Pow(s.temp[t, r], s.dfnl) * Math.Pow(ypc / ypc90, s.vbel);

                s.schisto[t, r] = s.smbs[r] * s.population[t, r] * s.smch[r] * Math.Pow(s.temp[t, r], s.smnl) * Math.Pow(ypc / ypc90, s.vbel);

                if (s.schisto[t, r] < -s.smbs[r] * s.population[t, r] * Math.Pow(ypc / ypc90, s.vbel))
                    s.schisto[t, r] = -s.smbs[r] * s.population[t, r] * Math.Pow(ypc / ypc90, s.vbel);

                s.malaria[t, r] = s.malbs[r] * s.population[t, r] * s.malch[r] * Math.Pow(s.temp[t, r], s.malnl) * Math.Pow(ypc / ypc90, s.vbel);
            }

        }
    }
}

File: src/components/ImpactVectorBorneDiseasesComponent.jl
@defcomp impactvectorbornediseases begin
    regions = Index()

    dengue = Variable(index=[time,regions])
    schisto = Variable(index=[time,regions])
    malaria = Variable(index=[time,regions])

    dfbs = Parameter(index=[regions])
    dfch = Parameter(index=[regions])
    smbs = Parameter(index=[regions])
    smch = Parameter(index=[regions])
    malbs = Parameter(index=[regions])
    malch = Parameter(index=[regions])

    gdp90 = Parameter(index=[regions])
    pop90 = Parameter(index=[regions])

    income = Parameter(index=[time,regions])
    population = Parameter(index=[time,regions])
    temp = Parameter(index=[time,regions])

    dfnl = Parameter(default=1)
    vbel = Parameter(default=-2.65)
    smnl = Parameter(default=1)
    malnl = Parameter(default=1)

    function run_timestep(p, v, d, t)

        for r in d.regions
            ypc = 1000.0 * p.income[t, r] / p.population[t, r]
            ypc90 = p.gdp90[r] / p.pop90[r] * 1000.0

            v.dengue[t, r] = p.dfbs[r] * p.population[t, r] * p.dfch[r] * p.temp[t, r]^p.dfnl * (ypc / ypc90)^p.vbel

            v.schisto[t, r] = p.smbs[r] * p.population[t, r] * p.smch[r] * p.temp[t, r]^p.smnl * (ypc / ypc90)^p.vbel

            if v.schisto[t, r] < -p.smbs[r] * p.population[t, r] * (ypc / ypc90)^p.vbel
                v.schisto[t, r] = -p.smbs[r] * p.population[t, r] * (ypc / ypc90)^p.vbel
            end

            v.malaria[t, r] = p.malbs[r] * p.population[t, r] * p.malch[r] * p.temp[t, r]^p.malnl * (ypc / ypc90)^p.vbel
        end
    end
end